pkg/dockerregistry/server/manifestservice.go

```diff
@@ -0,0 +1,279 @@
+package server
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+	"github.com/docker/distribution/manifest/schema2"
+	regapi "github.com/docker/distribution/registry/api/v2"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	quotautil "github.com/openshift/origin/pkg/quota/util"
+)
+
+var _ distribution.ManifestService = &manifestService{}
+
+type manifestService struct {
+	ctx       context.Context
+	repo      *repository
+	manifests distribution.ManifestService
+
+	// acceptschema2 allows to refuse the manifest schema version 2
+	acceptschema2 bool
+}
+
+// Exists returns true if the manifest specified by dgst exists.
+func (m *manifestService) Exists(ctx context.Context, dgst digest.Digest) (bool, error) {
+	context.GetLogger(ctx).Debugf("(*manifestService).Exists")
+
+	image, err := m.repo.getImage(dgst)
+	if err != nil {
+		return false, err
+	}
+	return image != nil, nil
+}
+
+// Get retrieves the manifest with digest `dgst`.
+func (m *manifestService) Get(ctx context.Context, dgst digest.Digest, options ...distribution.ManifestServiceOption) (distribution.Manifest, error) {
+	context.GetLogger(ctx).Debugf("(*manifestService).Get")
+
+	if _, err := m.repo.getImageStreamImage(dgst); err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStreamImage %s/%s@%s: %v", m.repo.namespace, m.repo.name, dgst.String(), err)
+		return nil, err
+	}
+
+	image, err := m.repo.getImage(dgst)
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving image %s: %v", dgst.String(), err)
+		return nil, err
+	}
+
+	ref := imageapi.DockerImageReference{
```

Comment:
@mfojtik  I don't like the idea of doing the function for these three assignments ... especially if the result changes just after call. Right now whole `DockerImageReference` creation is here.